students/luyao_xu/lesson04/mailroom2.py

```diff
@@ -0,0 +1,104 @@
+    donor.append(donation_amount)
+    while True:
+        more = input('Is there another gift amount you want to enter?(y/n) ')
+        if more[0].lower() == 'y':
+            donation_amount = float(input('That\'s so nice of you! Enter the amount you want to donate: $'))
+            # Once an amount has been given, add that amount to the donation history of the selected user
+            donor.append(donation_amount)
+        # Use string formatting to compose an email thanking the donor for their generous donation.
+        else:
+            print('Dear {}, we want to thank you for your total donation amount of ${}. Have a nice day!'
+                  .format(donor[0], sum(donor[1:])))
+            break
+    return donor
+
+
+def send_thank():
+    while True:
+        # If the user types ‘list’, show them a list of the donor names and re-prompt
+
+        user_input = input('Please enter a full name, type "list" to see the list of donors, type "back" to go back: ')
+        if user_input.lower() == 'list':
+            for name in donations:
+                print(name['name'])
+            continue
+
+        elif user_input.lower() == 'back':
+            # back to the main menu
+            return
+        # If the user types a name in the list, use it.
+        elif user_input.lower() in [d['name'].lower() for d in donations]:
+            for d in donations:
+                if d['name'] == user_input:
+                    donate(d)
+                    # exit for loop on name match
+                    break
+        else:
+            # If the user types a name not in the list, add that name to the data structure and use it
+            d = [user_input]
+            donations.append(donate(d))
+            break
+
+
+def sort_total_given(k):
+    return sum(k['donation amount'])
+
+
+def create_report():
+    donations.sort(key=sort_total_given, reverse=True)
+    #  print a list of your donors, sorted by total historical donation amount
+    head = '{:<26} | {:^11} | {:^8} | {:>12}'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+
+    print(head)
+    print('-' * len(head))
+    for i in donations:
+        name = i['name']
+        total_donation = sum(i['donation amount'])
+        gift_number = len(i['donation amount'])
+        average_donation = total_donation / gift_number
+
+        print('{:<27} ${:>11.2f} {:>12} ${:>13.2f}'.format(name, total_donation, gift_number, average_donation))
+
+
+def send_to_everyone():
+    letter = 'Dear {}, we want to thank you for your total donation amount of ${}. Have a nice day!'
+    for donation in donations:
+        with open(donation['name'] + '.txt', 'w') as outfile:
+            outfile.write(letter.format(donation['name'], sum(donation['donation amount'])))
+
+
+def menu_selection(prompt, dispatch_dict):
+    while True:
+        response = input(prompt)
+        result = dispatch_dict[response]()
+        if result == "exit":
+            break
```

Comment:
nice!

to make this even nicer you can map `quit_program` to exit action so you don't need to check for or except return value in your functions.

```
def quit_program():
    print("bye.")
    sys.exit()  # import sys at the top
```